File: proj/planerNEW/Volleyball/BaseGameHandling.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Volleyball
{
    [Serializable]
    class BaseGameHandling
    {
        #region members
        Logging log;
        int setCounter;
        int minutes;
        int pause;
        int[] rotationArray;
        public List<MatchData> matchData;
        public List<List<ResultData>> resultData;
        #endregion

        public BaseGameHandling(Logging log)
        {
            this.log = log;
            matchData = new List<MatchData>();
            resultData = new List<List<ResultData>>();
        }       

        public void setTimeParameters(int setCounter, int minutes, int pause)
        {
            this.setCounter = setCounter;
            this.minutes = minutes;
            this.pause = pause;
        }

        public void createRotationArray(int count)
        {
            rotationArray = new int[count];

            for (int i = 0; i < count; i++)
                rotationArray[i] = i + 1;
        }

        public void shiftRight(int[] list)
        {
            if (list.Length < 2)
                return;

            int oldFirst = list[0];

            for (int i = 0; i < (list.Length - 1); i++)
                list[i] = list[i + 1];

            list[list.Length - 1] = oldFirst;
        }

        // calculate results
        public void calculateResults()
        {
            List<TeamResult> teamResults = CalculateResults.addResultsForQualifyingAndInterimRounds(CalculateResults.calculateResults(matchData));

            foreach (TeamResult teamResult in teamResults)
            {
                foreach(List<ResultData> resultDataList in resultData)
                {
                    foreach(ResultData resultDataTeam in resultDataList)
                    {
                        if(resultDataTeam.Team == teamResult.TeamName)
                        {
                            resultDataTeam.PointsSets = teamResult.PointsSet;
                            resultDataTeam.PointsMatches = teamResult.PointsMatch;
                            break;
                        }
                    }
                }
            }
        }

        public void insertGameTime(DateTime startTournament)
        {
            int addtime = ((setCounter * minutes) + pause) * 60;

            for (int i = 0, lastRound = 0; i < matchData.Count; i++)
            {
                int roundValue = matchData[i].Round;

                if (lastRound < roundValue)
                {
                    lastRound = roundValue;
                    startTournament = startTournament.AddSeconds(addtime);
                }

                matchData[i].Time = startTournament;                
            }
        }

        //recalculate time schedule
        public void recalculateTimeSchedule(int changeIndex, DateTime gameTime)
        {
            int round = matchData[changeIndex].Round;
            int addTime = ((setCounter * minutes) + pause) * 60;
            
            for (int i = changeIndex; i < matchData.Count; i++)
            {
                if (round != matchData[i].Round)
                {
                    gameTime = gameTime.AddSeconds(addTime);
                    round++;
                }

                matchData[i].Time = gameTime;
            }
        }

        public void insertGameNumber()
        {
            for (int i = 0; i < matchData.Count; i++)
                matchData[i].Game = i + 1;
        }

        public void insertRoundNumber(int teamsCount, int fieldCount)
        {
            for (int i = 0, ii = 1, round = 1; i < matchData.Count; i++)
            {
                matchData[i].Round = round;

                if (ii < fieldCount)
                {
                    ii++;
                }
                else
                {
                    ii = 1;
                    round++;
                }
            }
        }

        // insert field numbers
        public void insertFieldnumbersAndFieldnames(int fieldCount, List<String> fieldNames)
        {
            int field = 0;

            createRotationArray(fieldCount);

            foreach(MatchData md in matchData)
            {
                md.FieldNumber = rotationArray[field];
                md.FieldName = fieldNames[rotationArray[field] - 1];

                if (field < rotationArray.Length - 1)
                {
                    field++;
                }
                else
                {
                    field = 0;
                    shiftRight(rotationArray);
                }
            }
        }       

        // generate divisions result lists
        public void fillResultLists(List<List<String>> divisionsList)
        {
            resultData.Clear();

            foreach (List<String> divisionList in divisionsList)
            {
                List<ResultData> teamList = new List<ResultData>();
                foreach (String team in divisionList)
                    teamList.Add(new ResultData(team));

                resultData.Add(teamList);
            }
        }

        // check equal division results
        public List<String> checkEqualDivisionResults()
        {
            List<String> result = new List<String>();

            for (int i = 0; i < resultData.Count; i++)
            {
                /*List<List<String>> getTeams = readDatabase("select distinct ms1.ms from " + resultTableName
                                                           + prefix + " ms1, (select ms, satz, punkte, intern from " + resultTableName
                                                           + prefix + ") ms2 where ms1.satz = ms2.satz and  ms1.punkte = ms2.punkte and ms1.intern = ms2.intern and ms1.ms != ms2.ms");

                if (getTeams.Count == 2)
                {
                    List<String> team1 = getTeams[0];
                    List<String> team2 = getTeams[1];
                    String gamenumber = readDatabase("SELECT spiel from " + round + " where ms_a = '"
                                                 + team1[0] + "' and ms_b = '" + team2[0] + "' or ms_a = '"
                                                 + team2[0] + "' and ms_b = '" + team1[0] + "'")[0][0];

                    result.Add("0");
                    result.Add(gamenumber);
                    result.Add(team1[0]);
                    result.Add(team2[0]);
                    return result;
                }
                else if (getTeams.Count > 2)
                {
                    List<String> teams = new List<String>();

                    teams.Add("1");

                    foreach (List<String> team in getTeams)
                        teams.Add(team[0]);

                    return teams;
                }*/
            }

            return result;
        }

        public void sortResults()
        {
            foreach(List<ResultData> rd in resultData)
            {
                if (rd != null)
                {
                    for (int i = rd.Count - 1; i > 0; i--)
                    {
                        for (int ii = 0; ii < i; ii++)
                        {
                            if (rd[ii].PointsSets < rd[ii + 1].PointsSets)
                            {
                                ResultData team = rd[ii];

                                rd[ii] = rd[ii + 1];
                                rd[ii + 1] = team;
                            }
                            else if (rd[ii].PointsSets == rd[ii + 1].PointsSets 
                                && rd[ii].PointsMatches < rd[ii + 1].PointsMatches)
                            {
                                ResultData team = rd[ii];

                                rd[ii] = rd[ii + 1];
                                rd[ii + 1] = team;
                            }
                            else if (rd[ii].PointsSets == rd[ii + 1].PointsSets 
                                && rd[ii].PointsMatches == rd[ii + 1].PointsMatches
                                && rd[ii].InternalRank < rd[ii + 1].InternalRank)
                            {
                                ResultData team = rd[ii];

                                rd[ii] = rd[ii + 1];
                                rd[ii + 1] = team;
                            }
                        }
                    }

                    for (int x = 0; x < rd.Count; x++)
                        rd[x].Rank = x + 1;
                }
            }
        }
    }
}

File: proj/planer/basegamehandling.cpp
#include "basegamehandling.h"

// base constructor
BaseGameHandling::BaseGameHandling(Database *db, QStringList *grPrefix)
{
    this->db = db;
    this->grPrefix = grPrefix;
}

// base destructor
BaseGameHandling::~BaseGameHandling()
{
}

// set time parameters
void BaseGameHandling::setTimeParameters(int satz, int min, int pause)
{
    this->satz = satz;
    this->min = min;
    this->pause = pause;
}

// clear all data
void BaseGameHandling::clearAllData(QStringList *tables)
{
    QStringList querys;

    for(int i = 0; i < tables->size(); i++)
        querys << "DELETE FROM " + tables->at(i);

    dbWrite(&querys);
}

// calculate results
void BaseGameHandling::calculateResult(QString round, QString resultTableName)
{
    emit logMessages("INFO:: calculating " + round + " results");
    QStringList execQuerys;
    QList<QStringList> vrGameResults = dbRead("SELECT spiel, ms_a, ms_b, satz1a, satz1b, satz2a, satz2b, satz3a, satz3b FROM " + round + " WHERE ms_a != '---' ORDER BY id ASC");
    QList<CalculateResults::teamResult> teamResults = CalculateResults::addResultsVrZw(CalculateResults::calculateResults(&vrGameResults));

    foreach(CalculateResults::teamResult tR, teamResults)
    {
        QString division;
        for(int i = 0; i < grPrefix->size();i ++)
        {
            QString prefix = grPrefix->at(i);

            if(dbRead("SELECT * FROM " + resultTableName + prefix + " WHERE ms = '" + tR.teamName + "'").count() > 0)
                division = prefix;
        }
        execQuerys << "UPDATE " + resultTableName + division + " SET punkte=" + QString::number(tR.sets) + ", satz=" + QString::number(tR.points) + " WHERE ms = '" + tR.teamName + "'";
    }

    dbWrite(&execQuerys);
}

//recalculate time schedule
void BaseGameHandling::recalculateTimeSchedule(QTableView *qtv, QSqlTableModel *model)
{
    QTime zeit = qtv->currentIndex().data().toTime();
    int addzeit = ((satz * min) + pause) * 60;
    int runde = model->data(model->index(qtv->currentIndex().row(), 1)).toInt();

    for(int i = qtv->currentIndex().row(); i <= model->rowCount(); i++)
    {
        if(runde != model->data(model->index(i, 1)).toInt())
        {
            zeit = zeit.addSecs(addzeit);
            runde++;
        }
        model->setData(model->index(i, 3), zeit.toString("hh:mm"));
    }
}

// insert field numbers
QStringList BaseGameHandling::insertFieldNr(QString round, int gameCount, int fieldCount)
{
    QStringList querys;

    for (int i = 1, field = 1; i <= gameCount; i++)
    {
        for(int x = 1, fieldHelp = field; x <= fieldCount; x++, fieldHelp++, i++)
        {
            querys << "UPDATE " + round + " SET feldnummer = " + QString::number(fieldHelp) + " WHERE id = " + QString::number(i);
            if(fieldHelp >= fieldCount)
                fieldHelp = 0;
        }

        i--;

        if(field < fieldCount)
            field++;
        else
            field = 1;
    }

    return querys;
}

// insert field names
QStringList BaseGameHandling::insertFieldNames(QString round, QStringList *fieldnames)
{
    QStringList querys;

    for(int i = 1; i <= fieldnames->size(); i++)
        querys << "UPDATE " + round + " SET feldname = '" + fieldnames->at(i-1) + "' WHERE feldnummer = " + QString::number(i);

    return querys;
}

// generate qualifying divisions result table
QStringList BaseGameHandling::generateResultTables(QString round, QList<QStringList> *divisionsList)
{
    QStringList querys;

    for(int i = 0, prefix = 0; i < divisionsList->size(); i++, prefix++)
    {
        QStringList division = divisionsList->at(i);
        QString group = grPrefix->at(prefix);

        for(int x = 0; x < division.size(); x++)
        {
            QString team = division.at(x);
            querys << "INSERT INTO " + round + group + " VALUES(" + QString::number(x) + ",'" + team + "',0,0,0,0)";
        }
    }

    return querys;
}

// check equal division results
QStringList BaseGameHandling::checkEqualDivisionResults(QString round, QString resultTableName)
{
    QStringList result;

    for(int i = 0; i < grPrefix->size(); i++)
    {
        QString prefix = grPrefix->at(i);

        QList<QStringList> getTeams = dbRead("select distinct ms1.ms from " + resultTableName
                                                   + prefix + " ms1, (select ms, satz, punkte, intern from " + resultTableName
                                                   + prefix + ") ms2 where ms1.satz = ms2.satz and  ms1.punkte = ms2.punkte and ms1.intern = ms2.intern and ms1.ms != ms2.ms");

        if(getTeams.count() == 2)
        {
            QStringList team1 = getTeams.at(0);
            QStringList team2 = getTeams.at(1);
            QString gamenr = db->read("SELECT spiel from " + round + " where ms_a = '"
                                         + team1.at(0) + "' and ms_b = '" + team2.at(0) + "' or ms_a = '"
                                         + team2.at(0) + "' and ms_b = '" + team1.at(0)+ "'").at(0).at(0);

            result.append("0");
            result.append(gamenr);
            result.append(team1.at(0));
            result.append(team2.at(0));
            return result;
        }
        else if(getTeams.count() > 2)
        {
            QStringList teams;

            teams << "1";

            foreach(QStringList team, getTeams)
                teams << team.at(0);

            return teams;
        }
    }

    return result;
}

// write to database with query list
void BaseGameHandling::dbWrite(QStringList *querys)
{
    for(int i = 0; i < querys->size(); i++)
        db->write(querys->at(i));
}

// write to database with single query
void BaseGameHandling::dbWrite(QString query)
{
    db->write(query);
}

// read prefix from list
QString BaseGameHandling::getPrefix(int index)
{
    return grPrefix->at(index);
}

// return prefix count
int BaseGameHandling::getPrefixCount()
{
    return grPrefix->size();
}

// parse int to string
QString BaseGameHandling::string(int val)
{
    return QString::number(val);
}

// read from database
QList<QStringList> BaseGameHandling::dbRead(QString query)
{
    return db->read(query);
}
